neuromation/cli/blob_storage.py

```diff
@@ -0,0 +1,368 @@
+from neuromation.api.blob_storage import BlobListing, BucketListing, PrefixListing
+from neuromation.api.file_filter import FileFilter
+from neuromation.api.url_utils import _extract_path
+
+from .const import EX_OSFILE
+from .formatters import (
+    BaseBlobFormatter,
+    LongBlobFormatter,
+    SimpleBlobFormatter,
+    create_storage_progress,
+    get_painter,
+)
+from .root import Root
+from .storage import calc_filters, filter_option
+from .utils import (
+    command,
+    group,
+    option,
+    pager_maybe,
+    parse_blob_or_file_resource,
+    parse_blob_resource,
+)
+
+
+log = logging.getLogger(__name__)
+
+
+@group(name="blob")
+def blob_storage() -> None:
+    """
+    Blob storage operations.
+    """
+
+
+BlobListings = Union[BucketListing, BlobListing, PrefixListing]
+
+
+@command()
+@click.argument("paths", nargs=-1)
+@option(
+    "--human-readable",
+    "-h",
+    is_flag=True,
+    help="with -l print human readable sizes (e.g., 2K, 540M).",
+)
+@option("-l", "format_long", is_flag=True, help="use a long listing format.")
+@option(
+    "--sort",
+    type=click.Choice(["name", "size", "time"]),
+    default="name",
+    help="sort by given field, default is name.",
+)
+@option(
+    "-r",
+    "--recursive",
+    is_flag=True,
+    help="List all keys under the URL path provided, not just 1 level depths.",
+)
+async def ls(
+    root: Root,
+    paths: Sequence[str],
+    human_readable: bool,
+    format_long: bool,
+    sort: str,
+    recursive: bool,
+) -> None:
+    """
+    List buckets or bucket contents.
+    """
+    uris = [parse_blob_resource(path, root) for path in paths]
+
+    blob = root.client.blob_storage
+
+    blob_listings: List[List[BlobListings]] = []
+
+    if not uris:
+        # List Buckets instead of blobs in bucket
+        listing: List[BlobListings] = []
+        blob_listings.append(listing)
+        for bucket in await blob.list_buckets():
+            listing.append(bucket)
+    else:
+        for uri in uris:
+            bucket_name, key = blob._extract_bucket_and_key(uri)
+
+            listing = []
+            blob_listings.append(listing)
+
+            for res in await blob.list_blobs(
+                bucket_name=bucket_name, prefix=key, recursive=recursive,
+            ):
+                listing.append(res)
+
+    formatter: BaseBlobFormatter
+    if format_long:
+        # Similar to `ls -l`
+        formatter = LongBlobFormatter(human_readable=human_readable, color=root.color)
+    else:
+        # Similar to `ls -1`, default for non-terminal on UNIX. We show full uris of
+        # blobs, thus column formatting does not work too well.
+        formatter = SimpleBlobFormatter(root.color)
+
+    if len(blob_listings) > 1:
+        buffer = []
+        for uri, listing in zip(uris, blob_listings):
+            buffer.append(click.style(str(uri), bold=True) + ":")
+            buffer.extend(formatter(listing))
+        pager_maybe("".join(buffer), root.tty, root.terminal_size)
```

Comment:
The first argument of `pager_maybe()` is an iterable of lines.
```suggestion
        pager_maybe(buffer, root.tty, root.terminal_size)
```